File: Spark/src/main/scala/com/tu/bdap/pagerank/PageRank_BDAP.scala
package com.tu.bdap.pagerank

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.graphx.Graph
import org.apache.spark.graphx.lib.PageRank

import com.tu.bdap.utils.DataSetLoader


/**
  * This algorithm computes the relevance of every node in a graph <br>
  * by using the PageRank formula
  */
object PageRank_BDAP {
  /**
    * Loads a graph from local disk or hdfs and calculates PageRank value for each vertex
    * @param args args[0] should contain path, args[1] is an integer identifying the dataset
    */
  def main(args: Array[String]): Unit = {


    //Start the Spark context
    val conf = new SparkConf()
      .setAppName("PageRank")
      //.setMaster("local")

  val sc = new SparkContext(conf)

    //Set NumIterations
    val numIterations = 10

    //Check input arguments
    if (args.length < 2) {
      System.err.println("Invalid input arguments")
      return
    }

    //Set input Arguments
    val dataSetPath = args(0)
    val dataSet = args(1).toInt

    //Load DataSet
    val edges = dataSet match {
      case 1 => DataSetLoader.loadUSA(sc, dataSetPath)
      case 2 => DataSetLoader.loadTwitter(sc, dataSetPath)
      case 3 => DataSetLoader.loadFriendster(sc, dataSetPath)
      case _ => null
    }
    //Check if DataSet could be loaded
    if (edges == null) {
      System.err.println("Could not load DataSet")
      return
    }

    //Create Graph from edges
    val graph = Graph.fromEdges(edges, "")

    //Run PageRank provided by GraphX
    PageRank.run(graph, numIterations, 0.15).vertices.collect()
  }
}

File: Flink/src/main/java/com/tu/bdap/pagerank/PageRank_BDAP.java
package com.tu.bdap.pagerank;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.DataSet;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.graph.Edge;
import org.apache.flink.graph.Graph;
import org.apache.flink.graph.library.PageRank;

import com.tu.bdap.utils.DataSetID;
import com.tu.bdap.utils.DatasetLoader;

//This algorithm computes the importance of every node in a graph
public class PageRank_BDAP {

	public static void main(String[] args) throws Exception {

		// Initialize DataSetPath and DataSet
		String dataSetPath = "";
		int dataSet = -1;

		try {
			dataSetPath = args[0];
			dataSet=Integer.parseInt(args[1]);
		} catch (Exception e) {
			e.printStackTrace();
		}

		if (dataSetPath == "" || dataSet == -1) {
			System.err.println("Wrong input parameters!");
			return;
		}

		// Iterations
		int numIterations = 11;

		// Create execution environment
		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

		// Initialize DataSet with GraphData
		DataSet<Edge<Integer, Double>> edges = null;

		// Read DataSet with GraphData using GraphLoader Class
		switch (dataSet) {
		case DataSetID.USA:
			edges = DatasetLoader.loadUsaInt(dataSetPath, env);
			break;
		case DataSetID.TWITTER:
			edges = DatasetLoader.loadTwitterInt(dataSetPath, env);
			break;
		case DataSetID.FRIENDSTER:
			edges = DatasetLoader.loadFriendsterInt(dataSetPath, env);
			break;
		}

		// Check if DataSet could be loaded!
		if (edges == null) {
			System.err.println("DataSet could not be loaded!");
			return;
		}

		// Create Graph from DataSet
		Graph<Integer, Double, Double> graph = Graph.fromDataSet(edges, new MapFunction<Integer, Double>() {
			@Override
			public Double map(Integer value) throws Exception {
				return 1.0;
			}
		}, env);

		// Run PageRank provided by Flink Gelly
		PageRank<Integer> pr = new PageRank<>(0.85, numIterations);
		pr.run(graph).collect();
	}

}